01-indoor-floor-level-tracker/web-app/src/pages/api/device-trackers.ts

```diff
@@ -15,10 +17,27 @@ function validateMethod(req: NextApiRequest, res: NextApiResponse) {
   return true;
 }
 
-async function performRequest() {
-  const app = services().getAppService();
+async function performRequest(authStringObj: CookieValueTypes) {
+  const appService = services().getAppService();
+
   try {
-    return await app.getDeviceTrackerData();
+    let authObj: AuthToken = {};
+    if (authStringObj === undefined) {
+      authObj = await appService.getAuthToken();
+      authStringObj = JSON.stringify(authObj);
+    }
+    if (typeof authStringObj === "string") {
+      const isAuthTokenValid = appService.checkAuthTokenValidity(authStringObj);
+      if (!isAuthTokenValid) {
+        authObj = await appService.getAuthToken();
+        authStringObj = JSON.stringify(authObj);
+      }
+
+      authObj = JSON.parse(authStringObj);
+      setCookie("authTokenObj", authStringObj);
+
+      return await appService.getDeviceTrackerData(authObj);
+    }
```

Comment:
Please, break this out into a few functions in order to: 
- DRY up the code that's currently duplicated ~5 times.
- Separate out the next-js specific code.
- Separate out the cookie-specific code (in case someone who's not server-less prefers to use a filesystem, global variable, redis etc.)
- Separate out any other separate concerns you notice along the way. Each function should do one thing and do it well. 

Those changes should make this code a lot easier to read, understand, review, test, etc.

Bonus points (overkill for our toy app, but very important in some customer apps):
- Consider adding encryption so the notehub auth token is never in plaintext in the client's browser.

Reply:
Code dried up:
* Added `auth-token.ts` file for reusable token generation / verification functions on client-side API requests. Now, all client API calls (update env vars, update device name, etc.), call this function first to make sure their auth tokens are in order.
* Added `cookieAuth.ts` file for reusable functions to handle fetching, storing and transforming auth tokens as cookies.
* Used reusable functions from first two files mentioned to dry up code for `index.tsx` and `settings.tsx` pages `getServerSideProps` functions.